File: 5-longest-palindromic-substring/5-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.length();
        int maxi = 0;
        string ans;
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        for (int g = 0; g < n; g++) {
            for (int i = 0, j = g; j < n; ++i, j++) {
                if (g == 0) {
                    dp[i][j] = true;
                    if (maxi < 1) {
                        maxi = 1;
                        ans = s[i];
                    }
                } else if (g == 1) {
                    if (s[i] == s[j]) {
                        dp[i][j] = true;
                        if (maxi < 2) {
                            maxi = 2;
                            ans = s.substr(i, g + 1);
                        }
                    }
                } else {
                    if (s[i] == s[j] and dp[i + 1][j - 1]) {
                        dp[i][j] = true;
                        if (maxi < g + 1) {
                            maxi = g + 1;
                            ans = s.substr(i, g + 1);
                        }
                    } else {
                        dp[i][j] = false;
                    }
                }
            }
        }
        return ans;
    }
};

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        int n = s.length();
        if (n < 2)
            return s;
        int odd, even, start = 0, end = 0, curLen;
        for (int i = 0; i < n; ++i) {
            odd = expandFromMiddle(s, i, i);
            even = expandFromMiddle(s, i, i + 1);
            curLen = Math.max(odd, even);
            if (curLen > end - start + 1) {
                end = i + (curLen / 2);
                start = i - ((curLen - 1) / 2);
            }
        }
        return s.substring(start, end + 1);
    }
    private int expandFromMiddle(String s, int left, int right) {
        int n = s.length();
        if (n < 2)
            return n;
        while (left >= 0 && right < n && s.charAt(left) == s.charAt(right)) {
            --left;
            ++right;
        }
        return right - left - 1;
    }
}